data/src/main/java/com/linkedin/data/Data.java

```diff
@@ -693,6 +720,29 @@ private static boolean reachable(DataComplex source, Object destination)
     return false;
   }
 
+  /**
+   * Check for cycles.
+   *
+   * <p>This checks if the given {@link DataComplex} is present in the given set of ancestor complexes. If yes,
+   * then it indicates that a cycle exists and we throw an exception. If not, we add it to the set.</p>
+   *
+   * @param dataComplex The current {@link DataComplex}
+   * @param ancestorSet The set containing all ancestors of the {@link DataComplex}
+   * @param pathList The list of paths encountered so far. Used to populate a useful exception message when a cycle is
+   *                 detected.
+   * @throws IOException If this {@link DataComplex} is present in the ancestor set indicating a cycle.
+   */
+  private static void checkForCyclesAndAdd(DataComplex dataComplex,
+      Set<DataComplex> ancestorSet, List<String> pathList) throws IOException
+  {
+    if (ancestorSet.contains(dataComplex))
+    {
+      throw new IOException("Cycle detected. Path: " + pathList);
```

Comment:
nit: This cycled situation does not sounds like an IOException, we probably want to throw exceptions more close to whats happening.
One question to ask is, if previously we don't check this cycling condition and now we are blocking them, would it cause any unexpected runtime issue we have to deal with?

Reply:
We earlier used to check for cycles when adding any DataComplex to a DataList/DataMap. This is probably a bad tradeoff since we penalize everyone on set. Instead we decided to skip it there (except for in assert mode) and do it here. This was discussed with @karthikbalasub